### PR TITLE
Tupek/differentiable block solver

### PR DESCRIPTION
**Pull request by tupek2**
A few ongoing fixes and improvements

- [x] Add differentiable block solver.  This block solver uses block linear solvers, but otherwise calls our existing nonlinear solver (e.g., Newton) over the entirely block system.  This allows monolithic nonlinear solves.
- [x] Remove duplication in test_solid_mechanics_state_advancer
- [x] Initial version of test_thermo_mechanics.cpp a coupled test that uses the new FieldStore as a prototyp
- [x] Update to latest gretl, which fixes an asan issue and also improves performance in some cases
- [x] Change the default NewtonLinesearch to use 10 cutbacks.  Small interface changes around this to ensure Newton still defaults back to 0 cutbacks.
- [x] Taylor series style for residual update in warm-start to better handle physics which start out of equilibrium

### CODE REVIEW

File: src/smith/differentiable_numerics/tests/test_field_state.cpp [hunk 1]
```diff
@@ -33,7 +34,8 @@ struct MeshFixture : public ::testing::Test {
     double width = 1.0;
     mesh_ = std::make_shared<smith::Mesh>(mfem::Mesh::MakeCartesian2D(2, 2, mfem_shape, true, length, width), MESHTAG,
                                           0, 0);
-    checkpointer_ = std::make_shared<gretl::DataStore>(5);
+
+    checkpointer_ = std::make_shared<gretl::DataStore>(std::make_unique<gretl::StrummWaltherCheckpointStrategy>(5));
```
Comment (white238) [inline src/smith/differentiable_numerics/tests/test_field_state.cpp:38]: Would there ever be a reason to swap checkpointing strategies?

File: src/smith/differentiable_numerics/tests/test_thermo_mechanics.cpp [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+ * @brief Compute Green's strain from the displacement gradient
+ */
+template <typename T, int dim>
+auto greenStrain(const tensor<T, dim, dim>& grad_u)
+{
+  return 0.5 * (grad_u + transpose(grad_u) + dot(transpose(grad_u), grad_u));
+}
+
+/// @brief Green-Saint Venant isotropic thermoelastic model
+struct GreenSaintVenantThermoelasticMaterial {
```
Comment (white238) [inline src/smith/differentiable_numerics/tests/test_thermo_mechanics.cpp:40]: Why not use the one in `src/smith/physics/materials/green_saint_venant_thermoelastic.hpp`? Is the State being empty important? If so please call out the reason in the above comment.

File: src/smith/differentiable_numerics/tests/test_thermo_mechanics.cpp [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+    type.unknown_index = static_cast<int>(num_unknowns_);
+    to_fields_index_[type.name] = fields_.size();
+    to_unknown_index_[type.name] = num_unknowns_;
+    FieldState new_field = smith::createFieldState<Space>(*data_store_, Space{}, type.name, mesh_->tag());
+    fields_.push_back(new_field);
+    ++num_unknowns_;
+    boundary_conditions_.push_back(
+        std::make_shared<DirichletBoundaryConditions>(mesh_->mfemParMesh(), new_field.get()->space()));
+    SLIC_ERROR_IF(num_unknowns_ != boundary_conditions_.size(),
+                  "Inconcistency between num unknowns and boundary condition size");
```
Comment (white238) [inline src/smith/differentiable_numerics/tests/test_thermo_mechanics.cpp]:
```suggestion
                  "Inconsistency between num unknowns and boundary condition size");
```

File: src/smith/differentiable_numerics/differentiable_solver.cpp [hunk 4]
```diff
@@ -266,18 +266,153 @@ std::vector<DifferentiableBlockSolver::FieldPtr> LinearDifferentiableBlockSolver
...
 }
 
+NonlinearDifferentiableBlockSolver::NonlinearDifferentiableBlockSolver(std::unique_ptr<EquationSolver> s)
+    : nonlinear_solver_(std::move(s))
+{
+}
+
+void NonlinearDifferentiableBlockSolver::completeSetup(const std::vector<FieldT>&)
+{
+  // eventually may need something like: initializeSolver(&nonlinear_solver_->preconditioner(), u);
```
Comment (white238) [inline src/smith/differentiable_numerics/differentiable_solver.cpp]:
```suggestion
  // TODO: eventually may need something like: initializeSolver(&nonlinear_solver_->preconditioner(), u);
```

File: src/smith/differentiable_numerics/differentiable_solver.cpp [hunk 5]
```diff
@@ -266,18 +266,153 @@ std::vector<DifferentiableBlockSolver::FieldPtr> LinearDifferentiableBlockSolver
...
+    block_u->GetBlock(row_i) = *u_guesses[static_cast<size_t>(row_i)];
+  }
+
+  auto block_r = std::make_unique<mfem::BlockVector>(block_offsets);
+
+  auto residual_op_ = std::make_unique<mfem_ext::StdFunctionOperator>(
+      block_u->Size(),
+      [&residual_funcs, num_rows, &u_guesses, &block_r](const mfem::Vector& u_, mfem::Vector& r_) {
+        const mfem::BlockVector* u = dynamic_cast<const mfem::BlockVector*>(&u_);
+        SLIC_ERROR_IF(!u, "Invalid u cast in block differentiable solver to a blocl vector");
```
Comment (white238) [inline src/smith/differentiable_numerics/differentiable_solver.cpp]:
```suggestion
        SLIC_ERROR_IF(!u, "Invalid u cast in block differentiable solver to a block vector");
```

File: src/smith/differentiable_numerics/nonlinear_solve.hpp [hunk 6]
```diff
@@ -24,6 +24,9 @@ class DifferentiableBlockSolver;
 class BoundaryConditionManager;
 class DirichletBoundaryConditions;
 
+/// @brief magic number for represending a field which is not an argument of the weak form.
```
Comment (white238) [inline src/smith/differentiable_numerics/nonlinear_solve.hpp]:
```suggestion
/// @brief magic number for representing a field which is not an argument of the weak form.
```

File: src/smith/differentiable_numerics/time_integration_rule.hpp [hunk 7]
```diff
@@ -34,27 +34,25 @@ struct BackwardEulerFirstOrderTimeIntegrationRule {
...
 };
 
-/// @brief Options for second order time integration methods
-enum class SecondOrderTimeIntegrationMethod
-{
-  IMPLICIT_NEWMARK,  /// implicit newmark discretization
-  QUASI_STATIC  /// quasi-static, specifies current field, velocity is central difference (for quasi-static artificial
-                /// viscosity), and acceleration is lagged for cases where it is set to some fixed value in time.
-};
+/// Altenative name for Backward Euler which makes sense when restricting what are typically second order odes,
```
Comment (white238) [inline src/smith/differentiable_numerics/time_integration_rule.hpp]:
```suggestion
/// Alternative name for Backward Euler which makes sense when restricting what are typically second order odes,
```

File: src/smith/physics/common.hpp [hunk 8]
```diff
@@ -22,18 +22,19 @@ struct TimeInfo {
   {
   }
 
-  /// @brief accessor for time
-  double time() const { return time_.first; }
+  /// @brief accessor for time previous time
+  double time_old() const { return time_.first; }
```
Comment (white238) [inline src/smith/physics/common.hpp]: delete?

File: src/smith/differentiable_numerics/time_integration_rule.hpp [hunk 9]
```diff
@@ -67,24 +65,58 @@ struct SecondOrderTimeIntegrationRule {
...
     auto dt = t.dt();
     return (4.0 / (dt * dt)) * (field_new - field_old) - (4.0 / dt) * velo_old - accel_old;
   }
+};
+
+/// @brief encodes rules for time discretizing second order odes (involving first and second time derivatives).
+/// When solving f(u, u_dot, u_dot_dot, t) = 0
+/// this class provides the current discrete approximation for u, u_dot, and u_dot_dot as a function of
+/// (u^{n+1},u^n,u_dot^n,u_dot_dot^n).
+struct QuasiStaticSecondOrderTimeIntegrationRule {
```
Comment (lihanyu97) [inline src/smith/differentiable_numerics/time_integration_rule.hpp]: In what situation do we need to use a QuasiStaticSecondOrder integration rule? In the packaged solid dynamics solver, we are always going to use something like Newmark right?

File: src/smith/differentiable_numerics/time_integration_rule.hpp [hunk 10]
```diff
@@ -34,27 +34,25 @@ struct BackwardEulerFirstOrderTimeIntegrationRule {
...
-{
-  IMPLICIT_NEWMARK,  /// implicit newmark discretization
-  QUASI_STATIC  /// quasi-static, specifies current field, velocity is central difference (for quasi-static artificial
-                /// viscosity), and acceleration is lagged for cases where it is set to some fixed value in time.
-};
+/// Altenative name for Backward Euler which makes sense when restricting what are typically second order odes,
+/// for example transient solid mechanics, to the quasi-static approximation.  It happens that the implementation is
+/// identical to backward-Euler applied to first order systems as we want to be able to capture current velocity
+/// dependencies.
+using QuasiStaticFirstOrderTimeIntegrationRule = BackwardEulerFirstOrderTimeIntegrationRule;
```
Comment (lihanyu97) [inline src/smith/differentiable_numerics/time_integration_rule.hpp:47]: Same questions as the second order integrator

File: src/smith/differentiable_numerics/nonlinear_solve.cpp [hunk 11]
```diff
@@ -342,12 +340,23 @@ std::vector<FieldState> block_solve(const std::vector<WeakForm*>& residual_evals
...
+    for (size_t row_i = 0; row_i < num_rows; ++row_i) {
+      FEFieldPtr primal_field_row_i = diagonal_fields[row_i];
+      applyBoundaryConditions(time_info.time(), bc_managers[row_i], primal_field_row_i, nullptr);
     }
 
     for (size_t row_i = 0; row_i < num_rows; ++row_i) {
       for (size_t col_j = 0; col_j < num_rows; ++col_j) {
-        input_fields[row_i][block_indices[row_i][col_j]] = diagonal_fields[col_j];
+        size_t prime_unknown_ij = block_indices[row_i][col_j];
+        if (prime_unknown_ij != invalid_block_index) {
```
Comment (lihanyu97) [inline src/smith/differentiable_numerics/nonlinear_solve.cpp:357]: Add a SLIC_ERROR_IF for prime_unknow_ij == invalid_block_index?

File: src/smith/differentiable_numerics/nonlinear_solve.cpp [hunk 12]
```diff
@@ -389,29 +396,37 @@ std::vector<FieldState> block_solve(const std::vector<WeakForm*>& residual_evals
         std::vector<double> tangent_weights(row_field_inputs.size(), 0.0);
         for (size_t col_j = 0; col_j < num_rows; ++col_j) {
           size_t field_index_to_diff = block_indices[row_i][col_j];
-          tangent_weights[field_index_to_diff] = 1.0;
-          auto jac_ij = residual_evals[row_i]->jacobian(time_info, shape_disp_ptr.get(),
-                                                        getConstFieldPointers(row_field_inputs), tangent_weights);
-          jacobians[row_i].emplace_back(std::move(jac_ij));
-          tangent_weights[field_index_to_diff] = 0.0;
+          if (field_index_to_diff != invalid_block_index) {
```
Comment (lihanyu97) [inline src/smith/differentiable_numerics/nonlinear_solve.cpp:399]: Same as the previous comment

File: src/smith/differentiable_numerics/tests/test_field_state.cpp (hunk 1 above)
Comment (tupek2) [reply src/smith/differentiable_numerics/tests/test_field_state.cpp]: At the moment they are kind of these just as research options.  I think in the long run, we'll always use the same one once its settled.

File: src/smith/differentiable_numerics/nonlinear_solve.cpp (hunk 11 above)
Comment (tupek2) [reply src/smith/differentiable_numerics/nonlinear_solve.cpp]: no, I put the invalid index into the map so I know to skip it.  so I expect some to be invalid block index.

File: src/smith/differentiable_numerics/nonlinear_solve.cpp (hunk 12 above)
Comment (tupek2) [reply src/smith/differentiable_numerics/nonlinear_solve.cpp]: same

Comment (tupek2) [reply src/smith/differentiable_numerics/nonlinear_solve.cpp]: invalid_block_index is a valid back for this to have.  that is why its there, to communicate there is no field argument in a particular weak form.

File: src/smith/differentiable_numerics/differentiable_solver.cpp [hunk 13]
```diff
@@ -266,18 +266,153 @@ std::vector<DifferentiableBlockSolver::FieldPtr> LinearDifferentiableBlockSolver
...
+  SMITH_MARK_FUNCTION;
+
+  int num_rows = static_cast<int>(u_bars.size());
+  SLIC_ERROR_IF(num_rows < 0, "Number of residual rows must be non-negative");
+
+  std::vector<DifferentiableBlockSolver::FieldPtr> u_duals(static_cast<size_t>(num_rows));
+  for (int row_i = 0; row_i < num_rows; ++row_i) {
+    u_duals[static_cast<size_t>(row_i)] = std::make_shared<DifferentiableBlockSolver::FieldT>(
+        u_bars[static_cast<size_t>(row_i)]->space(), "u_dual_" + std::to_string(row_i));
+  }
```
Comment (btalamini) [inline src/smith/differentiable_numerics/differentiable_solver.cpp:360]: In the future, we could make utilities for (block vector -> subvector) and the inverse and stick them in common place.

**Review by btalamini — APPROVED** [review]
I read through all of this, and it looks good to me! Sorry to make you wait for a no-feedback response.

**Comment by btalamini** [comment]
Pinging @white238 , @ebchin, @lihanyu97 . Any outstanding concerns?

**Review by white238 — APPROVED** [review]

**Review by lihanyu97 — APPROVED** [review]
